building/components/homeworld-admin-tools/src/virt.py

```diff
@@ -0,0 +1,421 @@
+        last_len = len(message)
+    print("server found".ljust(last_len, " "))
+
+
+def qemu_check_nested_virt():
+    if util.readfile("/sys/module/kvm_intel/parameters/nested").strip() != b"Y":
+        command.fail("nested virtualization not enabled")
+
+
+def auto_supervisor(ops: setup.Operations, install_iso, disk_gb=25, kill_file=None):
+    config = configuration.get_config()
+    supervisor = config.keyserver
+    ops.add_operation("install supervisor node (this may take several minutes)",
+                      lambda: qemu_install(supervisor.hostname, install_iso, "manual", disk_gb, "hide"),
+                      supervisor)
+    # TODO: better scoping for this file
+    alive = os.path.join(configuration.get_project(), "supervisor.live.%d" % os.getpid())
+    if kill_file is None:
+        kill = os.path.join(configuration.get_project(), "supervisor.kill.%d" % os.getpid())
+    else:
+        kill = kill_file
+    # TODO: annotations, so that this can be --dry-run'd
+    ops.add_operation("start up supervisor node", lambda: qemu_scan_ssh(supervisor.hostname, kill_path=kill))
+    ops.add_subcommand(seq.sequence_supervisor)
+    if kill_file is None:
+        def wait_for_stopped():
+            util.writefile(kill, b"stop\n")
+            # TODO: don't busywait
+            while os.path.exists(kill):
+                time.sleep(0.1)
+        ops.add_operation("stop supervisor node", wait_for_stopped)
+
+
+def auto_node(ops: setup.Operations, node: configuration.Node, install_iso, disk_gb=25, kill_file=None):
+    ops.add_operation("install node @HOST (this may take several minutes)",
+                      lambda: qemu_install(node.hostname, install_iso, "auto-admit", disk_gb, "hide"),
+                      node)
+    if kill_file is None:
+        kill = os.path.join(configuration.get_project(), "node-%s.kill.%d" % (node.hostname, os.getpid()))
+    else:
+        kill = kill_file
+    ops.add_operation("start up node @HOST", lambda: qemu_launch(node.hostname, kill_path=kill), node)
+    if kill_file is None:
+        def wait_for_stopped():
+            util.writefile(kill, b"stop\n")
+            # TODO: don't busywait
+            while os.path.exists(kill):
+                time.sleep(0.1)
+        ops.add_operation("stop supervisor node", wait_for_stopped)
+
+
+def auto_cluster(ops: setup.Operations, authorized_key=None):
+    if authorized_key is None:
+        # TODO: handle HOME being unset
+        authorized_key = os.path.join(os.getenv("HOME"), ".ssh/id_rsa.pub")
+    project, config = configuration.get_project(), configuration.get_config()
+    iso_path = os.path.join(project, "cluster-%d.iso" % os.getpid())
+    ops.add_operation("check nested virtualization", qemu_check_nested_virt)
+    ops.add_operation("update known hosts", access.update_known_hosts)
+    ops.add_operation("generate ISO", lambda: iso.gen_iso(iso_path, authorized_key, "serial"))
+    ops.add_operation("set up networking", net_up)
```

Comment:
I gave this a try. Unfortunately, net_up and net_down need to be able to be invoked separately (because we need to be able to 'spire virt down' at any time and have it do as much as possible), which is difficult to make easily workable with context managers.

Reply:
Could you explain a little further?  Could we have both a context manager and the ability to call `net_down` independently if needed?

Reply:
Nevermind. I realized I was looking at the problem wrong.